staging/src/k8s.io/kubectl/pkg/cmd/apply/apply.go

```diff
@@ -993,14 +1012,21 @@ func (o *ApplyOptions) MarkObjectVisited(info *resource.Info) error {
 func (o *ApplyOptions) PrintAndPrunePostProcessor() func() error {
 
 	return func() error {
+		ctx := context.TODO()
 		if err := o.printObjects(); err != nil {
 			return err
 		}
 
 		if o.Prune {
 			if cmdutil.ApplySet.IsEnabled() && o.ApplySet != nil {
-				p := newApplySetPruner(o)
-				return p.pruneAll()
+				pruner := newApplySetPruner(o)
+				if err := pruner.pruneAll(ctx, o.ApplySet); err != nil {
+					applySetErr := o.ApplySet.UpdateParent(SetUpdateIncomplete, o.DryRunStrategy, o.ValidationDirective)
```

Comment:
@justinsb this is an interesting case that I only really thought of when trying to come up with a testing strategy for this PR without yours under it. It didn't come up on the KEP--lmk what you think of the behaviour I'm proposing.

Reply:
So I think the case is "what if prune only partially succeeds"...

So before we apply I think we want to "increase" the annotations to cover the namespaces/gvks, so that they always cover any objects we apply.  After we apply & prune successfully we want to "decrease" the annotations so that we don't need to overfetch on the next apply/prune.  But if we can't "decrease" or can't prune that doesn't affect overall correctness, we just aren't done yet; the next time prune/apply is invoked we will just continue.

So if we fail to prune, I think we want to leave the overly-broad annotations.  Technically we could try to remove any GVKs we were able to successfully prune entirely, but I think that is going to be complicated and I don't see a lot of upside.

That's assuming I understood the case you were referring to correctly :-)

Reply:
If I understand the use case here correctly, I'm guessing we'll need more tests around interrupted in the middle of processing pruning, how will apply behave, can it properly continue after another invocation. IMO it should be able to easily figure out that the previous invocation only partially succeeded and continue warning about problems as it encounters them.

Reply:
@justinsb yes, this is about the case where pruning fails. Right now this PR doesn't patch the parent at the beginning of the apply, which is why I was needing to do it here even if the prune fails. Doing it earlier would indeed be more solid. It would mean an extra request being made unconditionally, but I think you're right that it is worth it. It should be straightforward to move this code to right after we've calculated the "updated" set.

@soltysh this PR does already have the foundations of that test! As part of `TestApplySetParentManagement`, we simulate failed apply and show that we end up with the correct state after the failed command. There is a TODO to add a segment to that test to show the automatic "repair" on the next apply being successful (which isn't possible to implement since successful pruning isn't implemented as of this PR).